packages/agents/chain-abstraction/src/helpers/swaputil.ts

```diff
@@ -0,0 +1,138 @@
+import { AlphaRouter } from "@uniswap/smart-order-router";
+import { ethers } from "ethers";
+import { CurrencyAmount, Token, TradeType } from "@uniswap/sdk-core";
+import { Route, Trade, Token as _Token, Fetcher, TokenAmount, Pair } from "@uniswap/sdk";
+import { FeeAmount } from "@uniswap/v3-sdk";
+import {
+  Fetcher as PancakeFetcher,
+  Trade as PancakeTrade,
+  Token as PancakeToken,
+  Route as PancakeRoute,
+  CurrencyAmount as PancakeCurrencyAmount,
+} from "@pancakeswap/sdk";
+
+type SwapPathCallBackArgs = {
+  fromTokenContractAddress: string;
+  toTokenContractAddress: string;
+  signerAddress: string;
+  chainId: number;
+  rpc: string;
+  amount: string;
+  fromTokenDecimal?: number;
+  toTokenDecimal?: number;
+};
+
+export type DestinationSwapPathCallback = (args: SwapPathCallBackArgs) => Promise<any>;
+
+const FEE_SIZE = 3;
+
+export function encodePath(path: string[], fees: FeeAmount[]): string {
+  if (path.length != fees.length + 1) {
+    throw new Error("path/fee lengths do not match");
+  }
+
+  let encoded = "0x";
+  for (let i = 0; i < fees.length; i++) {
+    // 20 byte encoding of the address
+    encoded += path[i].slice(2);
+    // 3 byte encoding of the fee
+    encoded += fees[i].toString(16).padStart(2 * FEE_SIZE, "0");
+  }
+  // encode the final token
+  encoded += path[path.length - 1].slice(2);
+
+  return encoded.toLowerCase();
+}
+
+/**
+ * Returns the `path` from the uniswap v3 router.
+ */
+export const getSwapPathForUniV3 = async (_args: SwapPathCallBackArgs) => {
+  try {
+    const { fromTokenContractAddress, toTokenContractAddress, chainId, rpc, fromTokenDecimal, toTokenDecimal, amount } =
+      _args;
+    const provider = new ethers.providers.JsonRpcProvider(rpc);
+    const tokenIn = new Token(chainId, fromTokenContractAddress, fromTokenDecimal ?? 18);
+    const tokenOut = new Token(chainId, toTokenContractAddress, toTokenDecimal ?? 18);
+    const amountIn = CurrencyAmount.fromRawAmount(tokenIn, +amount);
+
+    const router = new AlphaRouter({
+      chainId,
+      provider,
+    });
+
+    const routes = await router.route(amountIn, tokenOut, TradeType.EXACT_INPUT);
+
+    if (!routes) {
+      throw Error("No routes found for given Pair");
+    }
+
+    const tokenPathAddress = routes?.route[0].tokenPath.map((tokenPath) => tokenPath.address);
+
+    const feeAmount = [];
+
+    for (let i = 0; i < tokenPathAddress.length - 1; i++) {
+      feeAmount.push(FeeAmount.MEDIUM);
+    }
```

Comment:
This part seems like incorrect. because some pairs don't have medium fee tier. 

```typescript
const routes = await router.route(amountIn, tokenOut, TradeType.EXACT_INPUT);
```
this doesn't return the full path?

Reply:
It just returns the token path with address, doesn’t includes fees at all

Reply:
specially for the pools

Reply:
> It just returns the token path with address, doesn’t includes fees at all

why don't include fees? univ3 path should include fees as well. If not, swap will not work.

Reply:
I wondered about that as well. Seems little wierd